### ISSUE TITLE
[DOC] Document what resources are threadsafe

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/rmm#412]
**Describe the problems or issues found in the documentation**

Not all device resources are threadsafe and need to be used with the `thread_safe_resource_adaptor`. This should be documented more conspicuously.

**Comment by harrism** [comment]
Maybe it needs to be a trait of some sort.

**Comment by jrhemstad** [comment]
So the approach taken by `std::pmr` and `thrust::mr` is to indicate the thread safety in the name, e.g., [`std::pmr::(un)synchronized_pool_resource`](https://en.cppreference.com/w/cpp/header/memory_resource) and [`thrust::mr::(un)synchronized_pool_resource`](https://github.com/thrust/thrust/blob/master/thrust/mr/sync_pool.h).

So extending that to RMM...
```
cnmem_memory_resource -> synchronized_cnmem_resource
pool_memory_resource -> unsynchronized_pool_resource
thread_safe_adaptor<pool_memory_resource> -> synchronized_pool_resource
hybrid_memory_resource -> unsynchronized_hybrid_resource
thread_safe_adaptor<hybrid_memory_resource> -> synchronized_hybrid_resource
etc.
```

An alternative I've seen in other things like `spdlog` is that they'll use a `_st` or `_mt` suffix for single-threaded and multi-threaded, respectively. So:
```
cnmem_resource_mt
pool_resource_st
pool_resource_mt
```

I'd lean towards the former just for consistency with `std::pmr/thrust::mr`, but the latter option is definitely less verbose.

**Comment by harrism** [comment]
Naming conventions don't allow us to warn/assert in `thread_safe_resource_adaptor` when they try to adapt an already thread-safe resource.

**Comment by jrhemstad** [comment]
> Naming conventions don't allow us to warn/assert in `thread_safe_resource_adaptor` when they try to adapt an already thread-safe resource.

This would only work in a limited set of situations. First, it would require defining an explicit trait like `is_thread_safe` that is manually specialized for resources like `cnmem_memory_resource`, it wouldn't do any to guard against passing in a user-defined resource that is or is not threadsafe. 

Second, it would only work when the type of the upstream resource is known, e.g., `thread_safe_adaptor<cnmem_memory_resource>`. However, a common usecase is to type erase the upstream like `thread_safe_adaptor<device_memory_resource>`. We wouldn't be able to use a trait here because we don't know the derived type.

Long story short, I don't think we can programmatically enforce or warn about thread (un)safe resources, but we _can_ do our best to let users know what resources are (un)safe via good naming.

**Comment by harrism** [comment]
Maybe now that all of our memory resources are thread safe, we should take the stance that all memory resources are thread safe unless documented otherwise.
